pkg/hubble/parser/agent/parser.go

```diff
@@ -0,0 +1,255 @@
+		cidr = n.CIDR
+		identity = n.Identity
+		if n.OldIdentity != nil {
+			oldIdentity = &wrappers.UInt32Value{
+				Value: *n.OldIdentity,
+			}
+		}
+		if n.HostIP != nil {
+			hostIPString = n.HostIP.String()
+		}
+		if n.OldHostIP != nil {
+			oldHostIPString = n.OldHostIP.String()
+		}
+		encryptKey = n.EncryptKey
+		namespace = n.Namespace
+		podName = n.PodName
+	}
+	return &flowpb.AgentEvent{
+		Type: typ,
+		Notification: &flowpb.AgentEvent_IpcacheUpdate{
+			IpcacheUpdate: &flowpb.IPCacheNotification{
+				Cidr:        cidr,
+				Identity:    identity,
+				OldIdentity: oldIdentity,
+				HostIp:      hostIPString,
+				OldHostIp:   oldHostIPString,
+				EncryptKey:  uint32(encryptKey),
+				Namespace:   namespace,
+				PodName:     podName,
+			},
+		},
+	}
+}
+
+func notifyServiceUpsertedToProto(typ flowpb.AgentEventType, msg monitorAPI.AgentNotifyMessage) *flowpb.AgentEvent {
+	var id uint32
+	var feAddr *flowpb.ServiceUpsertNotificationAddr
+	var beAddrs []*flowpb.ServiceUpsertNotificationAddr
+	var svcType, svcTrafficPolicy, svcName, svcNamespace string
+	if n, ok := msg.Notification.(monitorAPI.ServiceUpsertNotification); ok {
+		id = n.ID
+		feAddr = &flowpb.ServiceUpsertNotificationAddr{
+			Ip:   n.Frontend.IP.String(),
+			Port: uint32(n.Frontend.Port),
+		}
+		for _, be := range n.Backends {
+			var ipStr string
+			if be.IP != nil {
+				ipStr = be.IP.String()
+			}
+			beAddrs = append(beAddrs, &flowpb.ServiceUpsertNotificationAddr{
+				Ip:   ipStr,
+				Port: uint32(be.Port),
+			})
+		}
+		svcType = n.Type
+		svcTrafficPolicy = n.TrafficPolicy
+		svcName = n.Name
+		svcNamespace = n.Namespace
+	}
+	return &flowpb.AgentEvent{
+		Type: typ,
+		Notification: &flowpb.AgentEvent_ServiceUpsert{
+			ServiceUpsert: &flowpb.ServiceUpsertNotification{
+				Id:               id,
+				FrontendAddress:  feAddr,
+				BackendAddresses: beAddrs,
+				Type:             svcType,
+				TrafficPolicy:    svcTrafficPolicy,
+				Name:             svcName,
+				Namespace:        svcNamespace,
+			},
+		},
+	}
+}
+
+func notifyServiceDeletedToProto(typ flowpb.AgentEventType, msg monitorAPI.AgentNotifyMessage) *flowpb.AgentEvent {
+	var id uint32
+	if n, ok := msg.Notification.(monitorAPI.ServiceDeleteNotification); ok {
+		id = n.ID
+	}
+	return &flowpb.AgentEvent{
+		Type: typ,
+		Notification: &flowpb.AgentEvent_ServiceDelete{
+			ServiceDelete: &flowpb.ServiceDeleteNotification{
+				Id: id,
+			},
+		},
+	}
+}
+
+func notifyUnknownToProto(typ flowpb.AgentEventType, msg monitorAPI.AgentNotifyMessage) *flowpb.AgentEvent {
+	n, _ := json.Marshal(msg.Notification)
+	return &flowpb.AgentEvent{
+		Type: typ,
+		Notification: &flowpb.AgentEvent_Unknown{
+			Unknown: &flowpb.AgentEventUnknown{
+				Type:         fmt.Sprintf("%d", msg.Type),
+				Notification: string(n),
+			},
+		},
+	}
+}
+
+func NotifyMessageToProto(msg monitorAPI.AgentNotifyMessage) *flowpb.AgentEvent {
+	switch msg.Type {
```

Comment:
Question: have you considered a type switch here on `msg.Notification` instead?

Since `msg` is only used to get its `Notification` member in the `*NotificationToProto` helper functions, we could pass the notification directly. With a type switch, we could avoid casting from all the helper functions:

```go
switch n := msg.Notification.(type) {
case monitorAPI.TimeNotification:
	if msg.Type == monitorAPI.AgentNotifyStart {
		return notifyTimeNotificationToProto(flowpb.AgentEventType_AGENT_STARTED, n)
	}
case monitorAPI.PolicyUpdateNotification:
	...
}
return notifyUnknownToProto(flowpb.AgentEventType_AGENT_EVENT_UNKNOWN, msg)
```
Would allow for

```go
func notifyTimeNotificationToProto(typ flowpb.AgentEventType, n monitorAPI.TimeNotification) *flowpb.AgentEvent`
```

Granted the switch will be uglier, but the `*NotificationToProto` may become obviously correct and feel "safer".

Reply:
Yeah, good point. I've considered this but went for the current implementation since for most `msg.Notification` members we still need to look at `msg.Type` (e.g. `monitorAPI.PolicyUpdateNotification` could be either be `monitorAPI.AgentNotifyPolicyUpdated` or `monitorAPI.AgentNotifyPolicyDeleted`) so went for a "cleaner" `NotifyMessageToProto`. But yeah, `*NotificationToProto` will be cleaner without the type switches. Given that I need to address https://github.com/cilium/cilium/pull/14168#issuecomment-734386525 anyway, I can give your proposal a go and compare again.

Reply:
Changed to use a type `switch`. The proposed solution indeed feels cleaner and safer. Thanks @kAworu!


<details>
  <summary>Incremental diff:</summary>

  ```diff
diff --git a/pkg/hubble/parser/agent/parser.go b/pkg/hubble/parser/agent/parser.go
index b5a1c1488f42..c5cd4f082a05 100644
--- a/pkg/hubble/parser/agent/parser.go
+++ b/pkg/hubble/parser/agent/parser.go
@@ -27,12 +27,10 @@ import (
 	"github.com/golang/protobuf/ptypes/wrappers"
 )
 
-func notifyTimeNotificationToProto(typ flowpb.AgentEventType, msg monitorAPI.AgentNotifyMessage) *flowpb.AgentEvent {
+func notifyTimeNotificationToProto(typ flowpb.AgentEventType, n monitorAPI.TimeNotification) *flowpb.AgentEvent {
 	var ts *timestamp.Timestamp
-	if notification, ok := msg.Notification.(monitorAPI.TimeNotification); ok {
-		if goTime, err := time.Parse(time.RFC3339Nano, notification.Time); err == nil {
-			ts, _ = ptypes.TimestampProto(goTime)
-		}
+	if goTime, err := time.Parse(time.RFC3339Nano, n.Time); err == nil {
+		ts, _ = ptypes.TimestampProto(goTime)
 	}
 	return &flowpb.AgentEvent{
 		Type: typ,
@@ -44,169 +42,116 @@ func notifyTimeNotificationToProto(typ flowpb.AgentEventType, msg monitorAPI.Age
 	}
 }
 
-func notifyPolicyNotificationToProto(typ flowpb.AgentEventType, msg monitorAPI.AgentNotifyMessage) *flowpb.AgentEvent {
-	var ruleCount int
-	var revision uint64
-	var labels []string
-	if n, ok := msg.Notification.(monitorAPI.PolicyUpdateNotification); ok {
-		ruleCount = n.RuleCount
-		labels = n.Labels
-		revision = n.Revision
-	}
-
+func notifyPolicyNotificationToProto(typ flowpb.AgentEventType, n monitorAPI.PolicyUpdateNotification) *flowpb.AgentEvent {
 	return &flowpb.AgentEvent{
 		Type: typ,
 		Notification: &flowpb.AgentEvent_PolicyUpdate{
 			PolicyUpdate: &flowpb.PolicyUpdateNotification{
-				RuleCount: int64(ruleCount),
-				Labels:    labels,
-				Revision:  revision,
+				RuleCount: int64(n.RuleCount),
+				Labels:    n.Labels,
+				Revision:  n.Revision,
 			},
 		},
 	}
 }
 
-func notifyEndpointRegenNotificationToProto(typ flowpb.AgentEventType, msg monitorAPI.AgentNotifyMessage) *flowpb.AgentEvent {
-	var id uint64
-	var labels []string
-	var e string
-	if n, ok := msg.Notification.(monitorAPI.EndpointRegenNotification); ok {
-		id = n.ID
-		labels = n.Labels
-		e = n.Error
-	}
-
+func notifyEndpointRegenNotificationToProto(typ flowpb.AgentEventType, n monitorAPI.EndpointRegenNotification) *flowpb.AgentEvent {
 	return &flowpb.AgentEvent{
 		Type: typ,
 		Notification: &flowpb.AgentEvent_EndpointRegenerate{
 			EndpointRegenerate: &flowpb.EndpointRegenNotification{
-				Id:     id,
-				Labels: labels,
-				Error:  e,
+				Id:     n.ID,
+				Labels: n.Labels,
+				Error:  n.Error,
 			},
 		},
 	}
 }
 
-func notifyEndpointUpdateNotificationToProto(typ flowpb.AgentEventType, msg monitorAPI.AgentNotifyMessage) *flowpb.AgentEvent {
-	var id uint64
-	var labels []string
-	var e, podName, namespace string
-	if n, ok := msg.Notification.(monitorAPI.EndpointNotification); ok {
-		id = n.ID
-		labels = n.Labels
-		e = n.Error
-		podName = n.PodName
-		namespace = n.Namespace
-	}
-
+func notifyEndpointUpdateNotificationToProto(typ flowpb.AgentEventType, n monitorAPI.EndpointNotification) *flowpb.AgentEvent {
 	return &flowpb.AgentEvent{
 		Type: typ,
 		Notification: &flowpb.AgentEvent_EndpointUpdate{
 			EndpointUpdate: &flowpb.EndpointUpdateNotification{
-				Id:        id,
-				Labels:    labels,
-				Error:     e,
-				PodName:   podName,
-				Namespace: namespace,
+				Id:        n.ID,
+				Labels:    n.Labels,
+				Error:     n.Error,
+				PodName:   n.PodName,
+				Namespace: n.Namespace,
 			},
 		},
 	}
 }
-func notifyIPCacheNotificationToProto(typ flowpb.AgentEventType, msg monitorAPI.AgentNotifyMessage) *flowpb.AgentEvent {
-	var cidr string
-	var identity uint32
+func notifyIPCacheNotificationToProto(typ flowpb.AgentEventType, n monitorAPI.IPCacheNotification) *flowpb.AgentEvent {
 	var oldIdentity *wrappers.UInt32Value
-	var hostIPString, oldHostIPString string
-	var encryptKey uint8
-	var namespace, podName string
-	if n, ok := msg.Notification.(monitorAPI.IPCacheNotification); ok {
-		cidr = n.CIDR
-		identity = n.Identity
-		if n.OldIdentity != nil {
-			oldIdentity = &wrappers.UInt32Value{
-				Value: *n.OldIdentity,
-			}
-		}
-		if n.HostIP != nil {
-			hostIPString = n.HostIP.String()
+	if n.OldIdentity != nil {
+		oldIdentity = &wrappers.UInt32Value{
+			Value: *n.OldIdentity,
 		}
-		if n.OldHostIP != nil {
-			oldHostIPString = n.OldHostIP.String()
-		}
-		encryptKey = n.EncryptKey
-		namespace = n.Namespace
-		podName = n.PodName
+	}
+	var hostIPString string
+	if n.HostIP != nil {
+		hostIPString = n.HostIP.String()
+	}
+	var oldHostIPString string
+	if n.OldHostIP != nil {
+		oldHostIPString = n.OldHostIP.String()
 	}
 	return &flowpb.AgentEvent{
 		Type: typ,
 		Notification: &flowpb.AgentEvent_IpcacheUpdate{
 			IpcacheUpdate: &flowpb.IPCacheNotification{
-				Cidr:        cidr,
-				Identity:    identity,
+				Cidr:        n.CIDR,
+				Identity:    n.Identity,
 				OldIdentity: oldIdentity,
 				HostIp:      hostIPString,
 				OldHostIp:   oldHostIPString,
-				EncryptKey:  uint32(encryptKey),
-				Namespace:   namespace,
-				PodName:     podName,
+				EncryptKey:  uint32(n.EncryptKey),
+				Namespace:   n.Namespace,
+				PodName:     n.PodName,
 			},
 		},
 	}
 }
 
-func notifyServiceUpsertedToProto(typ flowpb.AgentEventType, msg monitorAPI.AgentNotifyMessage) *flowpb.AgentEvent {
-	var id uint32
-	var feAddr *flowpb.ServiceUpsertNotificationAddr
-	var beAddrs []*flowpb.ServiceUpsertNotificationAddr
-	var svcType, svcTrafficPolicy, svcName, svcNamespace string
-	if n, ok := msg.Notification.(monitorAPI.ServiceUpsertNotification); ok {
-		id = n.ID
-		feAddr = &flowpb.ServiceUpsertNotificationAddr{
-			Ip:   n.Frontend.IP.String(),
-			Port: uint32(n.Frontend.Port),
-		}
-		for _, be := range n.Backends {
-			var ipStr string
-			if be.IP != nil {
-				ipStr = be.IP.String()
-			}
-			beAddrs = append(beAddrs, &flowpb.ServiceUpsertNotificationAddr{
-				Ip:   ipStr,
-				Port: uint32(be.Port),
-			})
+func notifyServiceUpsertedToProto(typ flowpb.AgentEventType, n monitorAPI.ServiceUpsertNotification) *flowpb.AgentEvent {
+	feAddr := &flowpb.ServiceUpsertNotificationAddr{
+		Ip:   n.Frontend.IP.String(),
+		Port: uint32(n.Frontend.Port),
+	}
+	beAddrs := make([]*flowpb.ServiceUpsertNotificationAddr, 0, len(n.Backends))
+	for _, be := range n.Backends {
+		var ipStr string
+		if be.IP != nil {
+			ipStr = be.IP.String()
 		}
-		svcType = n.Type
-		svcTrafficPolicy = n.TrafficPolicy
-		svcName = n.Name
-		svcNamespace = n.Namespace
+		beAddrs = append(beAddrs, &flowpb.ServiceUpsertNotificationAddr{
+			Ip:   ipStr,
+			Port: uint32(be.Port),
+		})
 	}
 	return &flowpb.AgentEvent{
 		Type: typ,
 		Notification: &flowpb.AgentEvent_ServiceUpsert{
 			ServiceUpsert: &flowpb.ServiceUpsertNotification{
-				Id:               id,
+				Id:               n.ID,
 				FrontendAddress:  feAddr,
 				BackendAddresses: beAddrs,
-				Type:             svcType,
-				TrafficPolicy:    svcTrafficPolicy,
-				Name:             svcName,
-				Namespace:        svcNamespace,
+				Type:             n.Type,
+				TrafficPolicy:    n.TrafficPolicy,
+				Name:             n.Name,
+				Namespace:        n.Namespace,
 			},
 		},
 	}
 }
 
-func notifyServiceDeletedToProto(typ flowpb.AgentEventType, msg monitorAPI.AgentNotifyMessage) *flowpb.AgentEvent {
-	var id uint32
-	if n, ok := msg.Notification.(monitorAPI.ServiceDeleteNotification); ok {
-		id = n.ID
-	}
+func notifyServiceDeletedToProto(typ flowpb.AgentEventType, n monitorAPI.ServiceDeleteNotification) *flowpb.AgentEvent {
 	return &flowpb.AgentEvent{
 		Type: typ,
 		Notification: &flowpb.AgentEvent_ServiceDelete{
 			ServiceDelete: &flowpb.ServiceDeleteNotification{
-				Id: id,
+				Id: n.ID,
 			},
 		},
 	}
@@ -226,30 +171,43 @@ func notifyUnknownToProto(typ flowpb.AgentEventType, msg monitorAPI.AgentNotifyM
 }
 
 func NotifyMessageToProto(msg monitorAPI.AgentNotifyMessage) *flowpb.AgentEvent {
-	switch msg.Type {
-	case monitorAPI.AgentNotifyStart:
-		return notifyTimeNotificationToProto(flowpb.AgentEventType_AGENT_STARTED, msg)
-	case monitorAPI.AgentNotifyPolicyUpdated:
-		return notifyPolicyNotificationToProto(flowpb.AgentEventType_POLICY_UPDATED, msg)
-	case monitorAPI.AgentNotifyPolicyDeleted:
-		return notifyPolicyNotificationToProto(flowpb.AgentEventType_POLICY_DELETED, msg)
-	case monitorAPI.AgentNotifyEndpointRegenerateSuccess:
-		return notifyEndpointRegenNotificationToProto(flowpb.AgentEventType_ENDPOINT_REGENERATE_SUCCESS, msg)
-	case monitorAPI.AgentNotifyEndpointRegenerateFail:
-		return notifyEndpointRegenNotificationToProto(flowpb.AgentEventType_ENDPOINT_REGENERATE_FAILURE, msg)
-	case monitorAPI.AgentNotifyEndpointCreated:
-		return notifyEndpointUpdateNotificationToProto(flowpb.AgentEventType_ENDPOINT_CREATED, msg)
-	case monitorAPI.AgentNotifyEndpointDeleted:
-		return notifyEndpointUpdateNotificationToProto(flowpb.AgentEventType_ENDPOINT_DELETED, msg)
-	case monitorAPI.AgentNotifyIPCacheUpserted:
-		return notifyIPCacheNotificationToProto(flowpb.AgentEventType_IPCACHE_UPSERTED, msg)
-	case monitorAPI.AgentNotifyIPCacheDeleted:
-		return notifyIPCacheNotificationToProto(flowpb.AgentEventType_IPCACHE_DELETED, msg)
-	case monitorAPI.AgentNotifyServiceUpserted:
-		return notifyServiceUpsertedToProto(flowpb.AgentEventType_SERVICE_UPSERTED, msg)
-	case monitorAPI.AgentNotifyServiceDeleted:
-		return notifyServiceDeletedToProto(flowpb.AgentEventType_SERVICE_DELETED, msg)
-	default:
-		return notifyUnknownToProto(flowpb.AgentEventType_AGENT_EVENT_UNKNOWN, msg)
+	switch n := msg.Notification.(type) {
+	case monitorAPI.TimeNotification:
+		if msg.Type == monitorAPI.AgentNotifyStart {
+			return notifyTimeNotificationToProto(flowpb.AgentEventType_AGENT_STARTED, n)
+		}
+	case monitorAPI.PolicyUpdateNotification:
+		if msg.Type == monitorAPI.AgentNotifyPolicyUpdated {
+			return notifyPolicyNotificationToProto(flowpb.AgentEventType_POLICY_UPDATED, n)
+		} else if msg.Type == monitorAPI.AgentNotifyPolicyDeleted {
+			return notifyPolicyNotificationToProto(flowpb.AgentEventType_POLICY_DELETED, n)
+		}
+	case monitorAPI.EndpointRegenNotification:
+		if msg.Type == monitorAPI.AgentNotifyEndpointRegenerateSuccess {
+			return notifyEndpointRegenNotificationToProto(flowpb.AgentEventType_ENDPOINT_REGENERATE_SUCCESS, n)
+		} else if msg.Type == monitorAPI.AgentNotifyEndpointRegenerateFail {
+			return notifyEndpointRegenNotificationToProto(flowpb.AgentEventType_ENDPOINT_REGENERATE_FAILURE, n)
+		}
+	case monitorAPI.EndpointNotification:
+		if msg.Type == monitorAPI.AgentNotifyEndpointCreated {
+			return notifyEndpointUpdateNotificationToProto(flowpb.AgentEventType_ENDPOINT_CREATED, n)
+		} else if msg.Type == monitorAPI.AgentNotifyEndpointDeleted {
+			return notifyEndpointUpdateNotificationToProto(flowpb.AgentEventType_ENDPOINT_DELETED, n)
+		}
+	case monitorAPI.IPCacheNotification:
+		if msg.Type == monitorAPI.AgentNotifyIPCacheUpserted {
+			return notifyIPCacheNotificationToProto(flowpb.AgentEventType_IPCACHE_UPSERTED, n)
+		} else if msg.Type == monitorAPI.AgentNotifyIPCacheDeleted {
+			return notifyIPCacheNotificationToProto(flowpb.AgentEventType_IPCACHE_DELETED, n)
+		}
+	case monitorAPI.ServiceUpsertNotification:
+		if msg.Type == monitorAPI.AgentNotifyServiceUpserted {
+			return notifyServiceUpsertedToProto(flowpb.AgentEventType_SERVICE_UPSERTED, n)
+		}
+	case monitorAPI.ServiceDeleteNotification:
+		if msg.Type == monitorAPI.AgentNotifyServiceDeleted {
+			return notifyServiceDeletedToProto(flowpb.AgentEventType_SERVICE_DELETED, n)
+		}
 	}
+	return notifyUnknownToProto(flowpb.AgentEventType_AGENT_EVENT_UNKNOWN, msg)
 }
diff --git a/pkg/hubble/parser/agent/parser_test.go b/pkg/hubble/parser/agent/parser_test.go
index 4ad7273ad8b9..0d97c99badef 100644
--- a/pkg/hubble/parser/agent/parser_test.go
+++ b/pkg/hubble/parser/agent/parser_test.go
@@ -75,7 +75,7 @@ func TestDecodeAgentEvent(t *testing.T) {
 		ev   *flowpb.AgentEvent
 	}{
 		{
-			name: "empty AgentNotifyMessage",
+			name: "empty",
 			msg:  api.AgentNotifyMessage{},
 			ev: &flowpb.AgentEvent{
 				Type: flowpb.AgentEventType_AGENT_EVENT_UNKNOWN,
@@ -88,7 +88,7 @@ func TestDecodeAgentEvent(t *testing.T) {
 			},
 		},
 		{
-			name: "unspecified AgentNotifyMessage",
+			name: "unspecified",
 			msg: api.AgentNotifyMessage{
 				Type:         api.AgentNotifyUnspec,
 				Notification: unknownNotification,
@@ -103,6 +103,23 @@ func TestDecodeAgentEvent(t *testing.T) {
 				},
 			},
 		},
+		{
+			name: "type and notification type mismatch",
+			msg: api.AgentNotifyMessage{
+				Type:         api.AgentNotifyStart,
+				Notification: unknownNotification,
+			},
+			ev: &flowpb.AgentEvent{
+				Type: flowpb.AgentEventType_AGENT_EVENT_UNKNOWN,
+				Notification: &flowpb.AgentEvent_Unknown{
+					Unknown: &flowpb.AgentEventUnknown{
+						Type:         fmt.Sprintf("%d", api.AgentNotifyStart),
+						Notification: string(unknownNotificationJSON),
+					},
+				},
+			},
+		},
+
 		{
 			name: "StartMessage",
 			msg:  api.StartMessage(agentStartTS),
```
</details>